## nn.Module로 구현하는 선형 회귀

- PyTorch에는 선형 회귀 모델, 평균 제곱오차가 구현되어 있다.   

### 1. 단순 선형 회귀 구현

In [3]:
import torch
import torch.nn as nn
import torch.nn.functional as F

torch.manual_seed(1)

In [4]:
x_train = torch.FloatTensor([[1], [2], [3]])
y_train = torch.FloatTensor([[2], [4], [6]])

In [5]:
# 선형 회귀 모델 nn.Linear(input_dim, output_dim)
model = nn.Linear(1, 1)   

In [6]:
# model에는 W와 b가 저장되어 있다. (랜덤 값)
print(list(model.parameters()))

[Parameter containing:
tensor([[0.5153]], requires_grad=True), Parameter containing:
tensor([-0.4414], requires_grad=True)]


In [7]:
# optimizer 정의
optimizer = torch.optim.SGD(model.parameters(), lr = 0.01)

In [8]:
EPOCH = 2000
for epoch in range(EPOCH+1):
    
    #H(x) 계산
    prediction = model(x_train)
    
    #cost 계산
    cost = F.mse_loss(prediction, y_train)
    
    #cost로 H(x) 개선하는 부분
    optimizer.zero_grad()
    cost.backward()
    optimizer.step()
    
    if epoch % 100 == 0:
        print('Epoch {:4d}/{} Cost : {:.6f}'.format(epoch, EPOCH, cost.item()))
    

Epoch    0/2000 Cost : 13.103541
Epoch  100/2000 Cost : 0.002791
Epoch  200/2000 Cost : 0.001724
Epoch  300/2000 Cost : 0.001066
Epoch  400/2000 Cost : 0.000658
Epoch  500/2000 Cost : 0.000407
Epoch  600/2000 Cost : 0.000251
Epoch  700/2000 Cost : 0.000155
Epoch  800/2000 Cost : 0.000096
Epoch  900/2000 Cost : 0.000059
Epoch 1000/2000 Cost : 0.000037
Epoch 1100/2000 Cost : 0.000023
Epoch 1200/2000 Cost : 0.000014
Epoch 1300/2000 Cost : 0.000009
Epoch 1400/2000 Cost : 0.000005
Epoch 1500/2000 Cost : 0.000003
Epoch 1600/2000 Cost : 0.000002
Epoch 1700/2000 Cost : 0.000001
Epoch 1800/2000 Cost : 0.000001
Epoch 1900/2000 Cost : 0.000000
Epoch 2000/2000 Cost : 0.000000


- W와 b가 최적화 되어있는지 확인해보자

In [12]:
new = torch.FloatTensor([[4.0]])
pred = model(new)
print(pred)

tensor([[7.9989]], grad_fn=<AddmmBackward0>)


- 8에 가까운걸 보아 어느정도 최적화된 것을 확인할 수 있다.

### 2. 다중 선형 회귀 구현

In [19]:
torch.manual_seed(1)

In [20]:
x_train = torch.FloatTensor([[73, 80, 75],
                            [93, 88, 93],
                            [89, 91, 90],
                            [96, 98, 100],
                            [73, 66, 70]])

y_train = torch.FloatTensor([[152], [185], [180], [196], [142]])

In [21]:
model = nn.Linear(3,1)

In [22]:
print(list(model.parameters()))

[Parameter containing:
tensor([[ 0.2975, -0.2548, -0.1119]], requires_grad=True), Parameter containing:
tensor([0.2710], requires_grad=True)]


In [23]:
optimizer = torch.optim.SGD(model.parameters(), lr=1e-5)

In [24]:
EPOCH = 2000

for epoch in range(EPOCH + 1):
    
    prediction = model(x_train)
    
    cost = F.mse_loss(prediction, y_train)
    
    optimizer.zero_grad()
    cost.backward()
    optimizer.step()
    
    if epoch % 100 == 0:
        print('Epoch {:4d}/{} Cost : {:.6f}'.format(epoch, EPOCH, cost.item()))

Epoch    0/2000 Cost : 31667.599609
Epoch  100/2000 Cost : 0.225988
Epoch  200/2000 Cost : 0.223909
Epoch  300/2000 Cost : 0.221935
Epoch  400/2000 Cost : 0.220059
Epoch  500/2000 Cost : 0.218270
Epoch  600/2000 Cost : 0.216571
Epoch  700/2000 Cost : 0.214954
Epoch  800/2000 Cost : 0.213418
Epoch  900/2000 Cost : 0.211954
Epoch 1000/2000 Cost : 0.210558
Epoch 1100/2000 Cost : 0.209230
Epoch 1200/2000 Cost : 0.207967
Epoch 1300/2000 Cost : 0.206767
Epoch 1400/2000 Cost : 0.205619
Epoch 1500/2000 Cost : 0.204521
Epoch 1600/2000 Cost : 0.203486
Epoch 1700/2000 Cost : 0.202486
Epoch 1800/2000 Cost : 0.201539
Epoch 1900/2000 Cost : 0.200635
Epoch 2000/2000 Cost : 0.199771


- W와 b 최적화 확인

In [25]:
new = torch.FloatTensor([[73, 80, 75]])
pred = model(new)
print(pred)

tensor([[151.2305]], grad_fn=<AddmmBackward0>)


- 당시 y는 152였는데 151로 예측했으므로 어느정도 최적화 되었다고 볼 수 있다.

In [26]:
print(list(model.parameters()))

[Parameter containing:
tensor([[0.9778, 0.4539, 0.5768]], requires_grad=True), Parameter containing:
tensor([0.2802], requires_grad=True)]
